lib/elements/clone.py

```diff
@@ -62,86 +51,87 @@ def clone(self):
+            # In a try block so we can ensure that the cloned_node is removed from the tree in the event of an exception.
+            # Otherwise, it might be left around on the document when this method returns.
 
-            angle_source = atan2(source_path[1][1][1] - source_path[0][1][1], source_path[1][1][0] - source_path[0][1][0])
-            angle_clone = atan2(clone_path[1][1][1] - clone_path[0][1][1], clone_path[1][1][0] - clone_path[0][1][0])
-            angle_embroidery = radians(-float(source_node.get(INKSTITCH_ATTRIBS['angle'], 0)))
+            cloned_node.set('transform', Transform(self.node.get('transform')) @ Transform(cloned_node.get('transform')))
 
-            diff = angle_source - angle_embroidery
-            rotation = degrees(diff + angle_clone)
+            flip_angle = self.flip_angle
 
-            if self.flip_angle:
-                rotation = -degrees(diff - angle_clone)
+            if self.clone_fill_angle is None:
+                # Calculate the rotation angle to apply to the fill of cloned elements, if not explicitly set
+                source_transform = source_node.composed_transform()
+                cloned_transform = cloned_node.composed_transform()
 
-        return -rotation
+                source_angle = angle_for_transform(source_transform)
+                cloned_angle = angle_for_transform(cloned_transform)
+                angle = cloned_angle - source_angle
 
-    def get_clone_style(self, style_name, node, default=None):
-        style = node.style[style_name] or default
-        return style
+                # Flip the angle if the clone is flipped relative to the source element
+                flip_angle = flip_angle ^ is_transform_flipped(source_transform) ^ is_transform_flipped(cloned_transform)
 
-    def center(self, source_node):
-        transform = get_node_transform(self.node.getparent())
-        center = self.node.bounding_box(transform).center
-        return center
+            elements = self.clone_to_elements(cloned_node)
+            for element in elements:
+                # We manipulate the element's node directly here instead of using get/set param methods, because otherwise
+                # we run into issues due to those methods' use of caching not updating if the underlying param value is changed.
+
+                if self.clone_fill_angle is None:  # Normally, rotate the cloned element's angle by the clone's rotation.
+                    element_angle = float(element.node.get(INKSTITCH_ATTRIBS['angle'], 0)) - angle
+                else:  # If clone_fill_angle is specified, override the angle instead.
+                    element_angle = self.clone_fill_angle
+
+                if flip_angle:
+                    element_angle = -element_angle
+
+                element.node.set(INKSTITCH_ATTRIBS['angle'], element_angle)
+
+                stitch_groups = element.to_stitch_groups(last_patch)
+                last_patch = stitch_groups[-1]
```

Comment:
I think we need to check if there are any stich groups before using an index. It is possible that to_stitch_groups returns an empty list.